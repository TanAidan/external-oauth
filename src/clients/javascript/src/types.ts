declare namespace Types {
  export interface ConnectOptions {
    authId?: string
    setupId?: string
    params?: any
  }

  export interface ConnectSuccess {
    authId: string
  }

  export interface ConnectError extends Error {}

  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
