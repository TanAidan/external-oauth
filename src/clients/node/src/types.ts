declare namespace Types {
  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
