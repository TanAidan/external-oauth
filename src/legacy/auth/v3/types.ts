import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
