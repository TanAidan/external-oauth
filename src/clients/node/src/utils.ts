import Types from './types'

/**
 * Helper to build a new URL from different params
 */
export function toURL(origin: string, baseURL: string, endpoint: string, queryString?: Types.RequestQueryString): URL {
  const removeLeadingSlash = (text: string) => {
    return text.replace(/^\//, '')
  }

  const removeTrailingSlash = (text: string) => {
    return text.replace(/\/$/, '')
  }

  const urlParts: string[] = []
  urlParts.push(removeTrailingSlash(origin))
  urlParts.push(removeLeadingSlash(removeTrailingSlash(baseURL)))
  urlParts.push(removeLeadingSlash(endpoint))

  const url = new URL(urlParts.join('/'))

  if (queryString) {
    Object.keys(queryString).forEach(key => url.searchParams.append(key, String(queryString[key])))
  }

  return url
}

/**
 * Helper to remove all undefined keys
 * @param obj {object}
 */
export function cleanHeaders(obj: Record<string, any>) {
  return Object.keys(obj).reduce((acc, key: string) => {
    if (obj[key] !== undefined) {
      acc[key] = obj[key]
    }
    return acc
  }, {} as Record<string, any>)
}
