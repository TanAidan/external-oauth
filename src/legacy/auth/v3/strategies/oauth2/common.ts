import jwt from 'jsonwebtoken'

import { TokenResult } from '../../../clients/oauth2'
import { IOAuth2Credentials } from '../../types'

export const NO_VALUE = 'non'

export const responseToCredentials = ({ accessToken, refreshToken, idToken, expiresIn }: TokenResult) => {
  const credentials: IOAuth2Credentials = {
    accessToken,
    refreshToken: refreshToken || NO_VALUE,
    idToken: idToken || NO_VALUE,
    expiresIn: expiresIn || 0
  }

  if (idToken) {
    credentials.idTokenJwt = getIdTokenJwt(idToken)
  }

  return credentials
}

export const getIdTokenJwt = (idToken?: string) => {
  if (!idToken) {
    return undefined
  }

  try {
    return jwt.decode(idToken) || undefined
  } catch (e) {
    return undefined
  }
}
