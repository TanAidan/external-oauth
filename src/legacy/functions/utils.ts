// Is the given object empty e.g. {}?
export function isEmpty(obj: Object) {
  return Object.getOwnPropertyNames(obj).length === 0
}
