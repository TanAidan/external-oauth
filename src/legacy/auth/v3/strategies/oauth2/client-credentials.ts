import { NextFunction, Response } from 'express'

import { TAuthenticateRequest } from '../../types'
import { getTokenWithClientCredentials } from '../../../clients/oauth2'
import { asyncMiddleware } from '../../../../errorHandler'
import { responseToCredentials } from './common'

export const authenticate = asyncMiddleware(async (req: TAuthenticateRequest, _res: Response, next: NextFunction) => {
  const { authorizationMethod, bodyFormat, config, tokenURL } = req.integrationConfig
  const { scope = [] } = config || {}
  const { clientId, clientSecret } = req.setupDetails.credentials

  const tokenResult = await getTokenWithClientCredentials({
    authorizationMethod,
    bodyFormat,
    clientId,
    clientSecret,
    scope,
    tokenURL
  })

  req.credentials = responseToCredentials(tokenResult)
  req.tokenResponse = tokenResult.decodedResponse

  next()
})
