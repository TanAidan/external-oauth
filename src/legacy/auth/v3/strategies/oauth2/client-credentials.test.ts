import { MiddlewareTestHarness } from '../../../../../../tests/utils'
import { TAuthenticateRequest, EAuthType } from '../../types'
import { authenticate } from './client-credentials'
// import { getTokenWithClientCredentials } from '../../../../clients/oauth2'
import { responseToCredentials } from './common'

jest.mock('../../../../clients/oauth2')
jest.mock('./common')

describe('authenticate', () => {
  const scope = ['print', 'email']
  const clientId = 'test-client-id'
  const clientSecret = 'test-client-secret'
  // const accessToken = 'test-access-token'
  // const refreshToken = 'test-refresh-token'
  // const idToken = 'test-id-token'
  // const expiresIn = 1234

  const tokenResult = { accessToken: 'response-access-token', decodedResponse: { body: { hey: 'there' } } }
  // const credentials = { accessToken, expiresIn, idToken, refreshToken }

  const setup = () =>
    new MiddlewareTestHarness<TAuthenticateRequest>({
      configureRequest: req => {
        req.integrationConfig = {
          authType: EAuthType.OAuth2,
          config: { scope },
          tokenURL: 'https://example.com/token'
        }
        req.setupDetails = { clientId, clientSecret }
      },
      testMiddleware: authenticate
    })

  beforeAll(() => {
    // mocked(getTokenWithClientCredentials).mockResolvedValue(tokenResult)
    // mocked(responseToCredentials).mockReturnValue(credentials)
  })

  beforeEach(() => {
    // mocked(responseToCredentials).mockClear()
  })

  it('fetches the tokens using the setup details', async () => {
    const test = setup()

    await test.get().expect(200)

    expect(responseToCredentials).toHaveBeenCalledWith(tokenResult)
    expect(test.req.credentials).toMatchSnapshot()
  })

  it('stores the token call response data on the request', async () => {
    const test = setup()

    await test
      .get()
      .query('code=test-code')
      .expect(200)

    expect(test.req.tokenResponse).toEqual(tokenResult.decodedResponse)
  })
})
