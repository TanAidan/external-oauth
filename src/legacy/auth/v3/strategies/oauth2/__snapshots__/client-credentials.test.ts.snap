// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate fetches the tokens using the setup details 1`] = `
Object {
  "accessToken": "test-access-token",
  "expiresIn": 1234,
  "idToken": "test-id-token",
  "refreshToken": "test-refresh-token",
}
`;
